tool/bindings_generator/translator.dart

```diff
@@ -80,19 +86,112 @@ class _Library {
   }
 }
 
-String _typeRaw(idl.IDLType idlType) {
-  final String type;
+(String, bool) _computeJsTypeUnion(
+    (String, bool) rawType1, (String, bool) rawType2) {
+  if (rawType1.$1 == rawType2.$1) {
+    return (rawType1.$1, rawType1.$2 || rawType2.$2);
+  }
+  // In the case of unions, we should try and get a JS type-able type to get a
+  // better LUB.
+  (String, bool) getTypeForUnionCalculation((String, bool) rawType) {
+    var (type, nullable) = rawType;
+    final decl = Translator.instance!._typeToDeclaration[type];
+    if (decl != null) {
+      final nodeType = decl.type.toDart;
+      switch (nodeType) {
+        case 'interface':
+        case 'dictionary':
+          // TODO(srujzs): We can do better if we do a LUB of the interfaces (so
+          // we get a possible interface name instead of `JSObject`), but that
+          // might be too much effort for too little reward. This would entail
+          // caching the hierarchy of IDL types.
+          type = 'JSObject';
+          break;
+        case 'typedef':
+          final desugared = _typeRaw((decl as idl.Typedef).idlType);
+          type = desugared.$1;
+          nullable = desugared.$2;
+          break;
+        case 'callback':
+          type = 'JSFunction';
+          break;
+        case 'enum':
+          type = 'JSString';
+          break;
+        default:
+          throw Exception('Unhandled type $type with node type: $nodeType');
+      }
+    }
+    return (type, nullable);
+  }
+
+  var (type1, nullable1) = getTypeForUnionCalculation(rawType1);
+  var (type2, nullable2) = getTypeForUnionCalculation(rawType2);
+  final nullable = nullable1 || nullable2;
+  if (!jsTypeHierarchy.containsKey(type1) ||
+      !jsTypeHierarchy.containsKey(type2)) {
+    return ('JSAny', nullable);
+  }
+  // Compute path from root and find the last type that exists in both paths for
+  // a least upper bound.
+  final t1Supertypes = [type1];
+  final t2Supertypes = [type2];
+  while (jsTypeHierarchy[type1]!.isNotEmpty) {
+    type1 = jsTypeHierarchy[type1]!.single;
+    t1Supertypes.add(type1);
+  }
+  while (jsTypeHierarchy[type2]!.isNotEmpty) {
+    type2 = jsTypeHierarchy[type2]!.single;
+    t2Supertypes.add(type2);
+  }
+  var t1i = t1Supertypes.length - 1;
+  var t2i = t2Supertypes.length - 1;
+  while (t1i >= 0 && t2i >= 0) {
+    if (t1Supertypes[t1i] != t2Supertypes[t2i]) break;
+    t1i--;
+    t2i--;
+  }
+  return (t1Supertypes[t1i + 1], nullable);
```

Comment:
I believe the logic is correct, but looking at this at first I was tempted to play through a few corner cases to feel more certain that this does what we intend it to do.

Not sure if it's worthwhile, but we could consider adding a small unit test to cover a couple corner cases (e.g. types near or at the root, equal types, strict subtypes, changing which type is t1 vs t2). I know equal types are handled early and don't hit this code path, but maybe we can pull out this part of the logic to test it separately (past where we computed the canonical type representation).

Reply:
I think that's fair. I've moved this code to a separate helper and added a test for it.

Re: equal types - equal "raw" types are handled early on. But if two different raw types map to the same *JS type*, we should handle that with this helper. I added a test for equal types to make sure.

A longer term goal for testing might be to get these scripts accepting a "fake" IDL for end-to-end testing, but that's probably a stretch goal. :)